File: bindings/zig/src/safe_network.zig
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>
//
// Proven SafeNetwork - FFI bindings to libproven network operations.
// All computation is performed in verified Idris 2 code via libproven.

const std = @import("std");
const c = @cImport(@cInclude("proven.h"));

/// Error types for network operations.
pub const NetworkError = error{
    InvalidIPv4,
    ProvenError,
};

/// IPv4 address (4 octets).
pub const IPv4 = struct {
    octets: [4]u8,

    /// Format as dotted-decimal string.
    pub fn format(self: IPv4, buf: []u8) ![]const u8 {
        return std.fmt.bufPrint(buf, "{d}.{d}.{d}.{d}", .{
            self.octets[0], self.octets[1], self.octets[2], self.octets[3],
        });
    }
};

/// Parse an IPv4 address string via libproven.
pub fn parseIPv4(address: []const u8) NetworkError!IPv4 {
    const result = c.proven_network_parse_ipv4(address.ptr, address.len);
    if (result.status != c.PROVEN_OK) return error.InvalidIPv4;
    return IPv4{ .octets = result.address.octets };
}

/// Check if an IPv4 address is private (RFC 1918) via libproven.
pub fn isPrivate(addr: IPv4) bool {
    const c_addr = c.ProvenIPv4Address{ .octets = addr.octets };
    return c.proven_network_ipv4_is_private(c_addr);
}

/// Check if an IPv4 address is a loopback address (127.0.0.0/8) via libproven.
pub fn isLoopback(addr: IPv4) bool {
    const c_addr = c.ProvenIPv4Address{ .octets = addr.octets };
    return c.proven_network_ipv4_is_loopback(c_addr);
}

/// Convenience: parse and check if private.
pub fn isPrivateStr(address: []const u8) bool {
    const addr = parseIPv4(address) catch return false;
    return isPrivate(addr);
}

/// Convenience: parse and check if loopback.
pub fn isLoopbackStr(address: []const u8) bool {
    const addr = parseIPv4(address) catch return false;
    return isLoopback(addr);
}

test "parseIPv4" {
    const ip = try parseIPv4("192.168.1.1");
    try std.testing.expectEqual(@as(u8, 192), ip.octets[0]);
    try std.testing.expectEqual(@as(u8, 168), ip.octets[1]);
    try std.testing.expectError(error.InvalidIPv4, parseIPv4("invalid"));
}

test "isPrivate" {
    try std.testing.expect(isPrivateStr("192.168.1.1"));
    try std.testing.expect(isPrivateStr("10.0.0.1"));
    try std.testing.expect(!isPrivateStr("8.8.8.8"));
}

test "isLoopback" {
    try std.testing.expect(isLoopbackStr("127.0.0.1"));
    try std.testing.expect(!isLoopbackStr("192.168.1.1"));
}

File: bindings/rust/src/safe_network.rs
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>

//! Safe network address operations via libproven FFI.
//!
//! Provides IPv4 address parsing and classification.
//! All operations delegate to Idris 2 verified code.

use crate::core::{self, Result};
use crate::ffi;

/// Parsed IPv4 address.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct Ipv4Addr {
    /// Four octets of the IPv4 address.
    pub octets: [u8; 4],
}

impl std::fmt::Display for Ipv4Addr {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{}.{}.{}.{}",
            self.octets[0], self.octets[1], self.octets[2], self.octets[3]
        )
    }
}

/// Safe network address operations.
pub struct SafeNetwork;

impl SafeNetwork {
    /// Parse an IPv4 address string.
    ///
    /// Returns `Err` if the string is not a valid dotted-decimal IPv4 address.
    pub fn parse_ipv4(input: &str) -> Result<Ipv4Addr> {
        let bytes = input.as_bytes();
        // SAFETY: We pass a valid pointer and length from the string's bytes.
        // The FFI function only reads from the pointer within the given length.
        let result = unsafe {
            ffi::proven_network_parse_ipv4(bytes.as_ptr(), bytes.len())
        };
        core::status_to_result(result.status)?;
        Ok(Ipv4Addr {
            octets: result.address.octets,
        })
    }

    /// Check if an IPv4 address is in a private range (RFC 1918).
    ///
    /// Private ranges: 10.0.0.0/8, 172.16.0.0/12, 192.168.0.0/16.
    pub fn ipv4_is_private(addr: &Ipv4Addr) -> bool {
        let ffi_addr = ffi::IPv4Address {
            octets: addr.octets,
        };
        // SAFETY: proven_network_ipv4_is_private takes a value-type struct
        // with no pointers; always safe to call.
        unsafe { ffi::proven_network_ipv4_is_private(ffi_addr) }
    }

    /// Check if an IPv4 address is a loopback address (127.0.0.0/8).
    pub fn ipv4_is_loopback(addr: &Ipv4Addr) -> bool {
        let ffi_addr = ffi::IPv4Address {
            octets: addr.octets,
        };
        // SAFETY: proven_network_ipv4_is_loopback takes a value-type struct
        // with no pointers; always safe to call.
        unsafe { ffi::proven_network_ipv4_is_loopback(ffi_addr) }
    }
}
